zeroconf/_protocol.py

```diff
@@ -75,6 +82,65 @@ def truncated(self) -> bool:
         return (self.flags & _FLAGS_TC) == _FLAGS_TC
 
 
+class DnsNameReader:
+    """Read dns domain names from a packet."""
+
+    def __init__(self, packet: bytes) -> None:
+        """Init the record with the packet."""
+        self.packet = packet
+        self.packet_len = len(packet)
+        self.name_cache: Dict[int, List[str]] = {}
+        self.seen: Set[int] = set()
+
+    def read_name_at_offset(self, offset: int) -> Tuple[int, str]:
+        """Read a dns domain name at an offset in the packet."""
+        labels: List[str] = []
+        self.seen.clear()
+        new_offset = self._decode_labels_at_offset(offset, labels)
+        labels.append("")
+        name = ".".join(labels)
+        if len(name) > MAX_NAME_LENGTH:
+            raise IncomingDecodeError(f"DNS name {name} exceeds maximum length of {MAX_NAME_LENGTH}")
+        return new_offset, name
+
+    def _decode_labels_at_offset(self, off: int, labels: List[str]) -> int:
+        # This is a tight loop that is called frequently, small optimizations can make a difference.
+        while off < self.packet_len:
+            length = self.packet[off]
+            if length == 0:
+                return off + DNS_COMPRESSION_HEADER_LEN
+
+            if length < 0x40:
+                label_idx = off + DNS_COMPRESSION_HEADER_LEN
+                labels.append(str(self.packet[label_idx : label_idx + length], 'utf-8', 'replace'))
+                off += DNS_COMPRESSION_HEADER_LEN + length
+                continue
+
+            if length < 0xC0:
+                raise IncomingDecodeError(f"DNS compression type {length} is unknown")
+
+            # We have a DNS compression pointer
+            link = (length & 0x3F) * 256 + self.packet[off + 1]
+            if link > self.packet_len:
+                raise IncomingDecodeError(f"DNS compression pointer at {off} points to {link} beyond packet")
+            if link == off:
+                raise IncomingDecodeError(f"DNS compression pointer at {off} points to itself")
+            if link in self.seen:
+                raise IncomingDecodeError(f"DNS compression pointer at {off} was seen again")
+            self.seen.add(link)
+            linked_labels = self.name_cache.get(link, [])
+            if not linked_labels:
+                self._decode_labels_at_offset(link, linked_labels)
+                self.name_cache[link] = linked_labels
+            labels.extend(linked_labels)
+            if len(labels) > MAX_DNS_LABELS:
+                raise IncomingDecodeError(f"Maximum dns labels reached while processing pointer at {off}")
+
+            return off + DNS_COMPRESSION_POINTER_LEN
```

Comment:
Probably going to continue here in case there is another label until we get a zero. In practice is probably never happens though￼

Reply:
That actually cannot happen.